### PR TITLE
OGIP 15 (Sammelaufgaben)

### PR DESCRIPTION
**Pull request by maethu**
Goal:
It should be possible to add multiple task with the add task form. For every responsible user there should be one separate task.

[Link to OGIP]

Tasks:
- [x] Add New `All` option to the add/edit task form.
- [x] New default value for the client field while adding tasks is `All`.
- [x] Make sure the new `All` option  is only available on tasks (add/edit)
- [x] Add multiple tasks if there's more than one user. Use plone dexterity form api.
- [x] Implement a new select2 based widget. #2910
- [x] Replace other gever specific autocomplete selection fields #2932 
- [x] Changelog

Styling: https://github.com/4teamwork/plonetheme.teamraum/pull/550

### CODE REVIEW

**Comment by maethu** [comment]
FYI @phgross @deiferni @lukasgraf

File: opengever/task/form.py [hunk 1]
```diff
@@ -48,9 +56,33 @@ def update(self):
                 u"help_responsible_single_client_setup", default=u"")
 
     def createAndAdd(self, data):
-        update_reponsible_field_data(data)
+        created = []
+        if isinstance(data['responsible'], basestring):
+            data['responsible'] = [data['responsible']]
+
+        for responsble in data['responsible']:
```
Comment (phgross) [inline opengever/task/form.py]: typo: responsble

File: opengever/task/tests/test_task.py [hunk 2]
```diff
@@ -269,6 +269,60 @@ def test_responsible_client_for_multiple_orgunits(self, browser):
...
+        browser.login().visit(dossier)
+        factoriesmenu.add('Task')
+        browser.fill({'Title': 'Task title',
+                      'Task Type': 'To comment',
+                      'Responsible': responsible_users})
+        browser.find('Save').click()
+
+        tasks = dossier.objectValues()
+        self.assertEquals(2, len(tasks), 'Expect 2 tasks')
+        self.assertEquals(TEST_USER_ID, tasks[0].responsible)
```
Comment (phgross) [inline opengever/task/tests/test_task.py]: I would add also an assertion on the `responsible_client` value.

File: opengever/task/tests/test_task.py [hunk 3]
```diff
@@ -269,6 +269,60 @@ def test_responsible_client_for_multiple_orgunits(self, browser):
...
+        self.assertEquals(user.userid, tasks[1].responsible)
```
Comment (phgross) [inline opengever/task/tests/test_task.py]: I would like to have a tests which checks if all the created tasks are correctly stored in the Globlaindex DB.

File: opengever/task/form.py [hunk 4]
```diff
@@ -87,6 +87,17 @@ def _create_task(self, task_payload):
         activity.record()
         return task
 
+    def _set_immediate_view(self):
```
Comment (phgross) [inline opengever/task/form.py]: Would it be possible to add a separate "redirection" when only a single task gets added than when multiple tasks are added? So we wouldn't change the behavior for the already existing functionality.

File: opengever/task/tests/test_task.py (hunk 2 above)
Comment (maethu) [inline opengever/task/tests/test_task.py]: 👍

File: opengever/task/tests/test_task.py [hunk 5]
```diff
@@ -213,6 +214,46 @@ def test_adding_a_document_add_response_on_main_task(self):
...
+
+        field = browser.css('[data-fieldname="form.widgets.responsible_client"]')
+        self.assertFalse(
+            field.css('input.hidden-widget'),
+            'There are multiple clients, the responsible_client shoud be '
+            'visible.')
+
+        self.assertEquals('ALL_ORGUNITS',
+                          field.css('select').first.value,
+                          'The default value should be the client of the user'
```
Comment (maethu) [inline opengever/task/tests/test_task.py]: 👍

File: opengever/task/tests/test_task.py (hunk 3 above)
Comment (maethu) [inline opengever/task/tests/test_task.py]: 👍

File: opengever/task/form.py (hunk 4 above)
Comment (maethu) [inline opengever/task/form.py]: 👍

File: opengever/task/form.py (hunk 1 above)
Comment (maethu) [inline opengever/task/form.py]: 👍

**Comment by maethu** [comment]
@phgross I changed/fixed the mentioned points. I now going ahead and implement the new ftw.keywordwidget with async option --> https://github.com/4teamwork/ftw.keywordwidget/pull/14

File: docs/HISTORY.txt [hunk 6]
```diff
@@ -50,6 +50,7 @@ Changelog
 - Support simple css colornames in the colorization viewlet. [phgross]
 - Dossierdetails PDF: Fix indentation of dossier metadata table. [lgraf]
 - Move the attach to email action to second position, after the copy action. [elioschmutz]
+- OGIP 15: Implement adding multiple task with the add form. [mathias.leimgruber]
```
Comment (phgross) [inline docs/HISTORY.txt]: Wrong release ⬆️

File: opengever/ogds/base/sources.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+@implementer(IQuerySource)
+class AllUsersAndInboxesSource(object):
+    """This example of a IQuerySource is taken from the
+    plone.formwidget.autocomplete
+    """
+
+    def __init__(self, context):
+        self.context = context
+        # self.vocab = SimpleVocabulary.fromItems(
```
Comment (phgross) [inline opengever/ogds/base/sources.py]: ✂️

File: opengever/ogds/base/tests/test_sources.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+        self.source = AllUsersAndInboxesSource(self.portal)
+
+    def test_get_term_by_token(self):
+        term = self.source.getTermByToken(u'unit1:hans')
+        self.assertTrue(isinstance(term, SimpleTerm))
+        self.assertEquals(u'Informatik: Peter Hans (test@example.org)',
+                          term.title)
+
+    def test_get_term_by_token_raises_BadReques_if_no_token(self):
```
Comment (phgross) [inline opengever/ogds/base/tests/test_sources.py]: Typo: BadReques*t*

File: sources.cfg [hunk 9]
```diff
@@ -5,6 +5,7 @@ extensions = mr.developer
 development-packages =
   opengever.maintenance
   collective.js.timeago
+  plonetheme.teamraum
```
Comment (phgross) [inline sources.cfg]: If there is a plonetheme.teamraum change necessary, please add the PR.

File: opengever/ogds/base/sources.py [hunk 10]
```diff
@@ -91,6 +95,25 @@ def search(self, query_string):
                 self.getTerm(u'{}:{}'.format(orgunit.id(), user.userid)))
         return self.terms
 
+    def get_client_id(self):
+        """Tries to get the client from the request. If no client is found None
+        is returned.
+        """
+
+        request = getRequest()
+        client_id = request.get('client',
```
Comment (phgross) [inline opengever/ogds/base/sources.py]: As far as I see `request.get('client',` should never be the case and is obsolete ...

File: opengever/ogds/base/tests/test_sources.py [hunk 11]
```diff
@@ -117,3 +117,21 @@ def test_getTerm_can_handle_values_containing_only_a_userid(self):
...
+    def test_getTerm_handles_inactive_users(self):
+        create(Builder('ogds_user')
+               .id('peter.muster')
+               .having(firstname='Peter', lastname='Muster', active=False)
+               .assign_to_org_units([self.org_unit2]))
+
+        self.assertTrue(self.source.getTerm('unit2:peter.muster'),
+                        'Expect a term from inactive user')
+
+    def test_search_for_nactive_users_is_not_possible(self):
```
Comment (phgross) [inline opengever/ogds/base/tests/test_sources.py]: Typo: test_search_for_nactive_users_is_not_possible --> test_search_for_**i**nactive_users_is_not_possible

File: opengever/ogds/base/sources.py [hunk 12]
```diff
@@ -44,13 +47,27 @@ def __len__(self):
...
         if len(data) == 2:
             orgunit_id, userid = data
         else:
             userid = value
             orgunit_id = self.client_id
 
+        # Handle special case - Inboxes: in form "inbox:unit_id"
+        if ActorLookup(value).is_inbox():
+            orgunit_id = userid
+            orgunit = OrgUnit.query.filter(OrgUnit.unit_id == orgunit_id).one()
```
Comment (phgross) [inline opengever/ogds/base/sources.py]: `orgunit = OrgUnit.query.get(orgunit_id)`

File: opengever/task/browser/assign.py [hunk 13]
```diff
@@ -130,9 +122,9 @@ def updateWidgets(self):
         super(AssignTaskForm, self).updateWidgets()
 
         # Prefill the responsible_client based on forwarding
-        fwd = self.context
-        self.widgets['responsible_client'].value = [fwd.responsible_client]
-        self.widgets['responsible_client'].mode = HIDDEN_MODE
+        # fwd = self.context
```
Comment (phgross) [inline opengever/task/browser/assign.py]: ✂️

File: opengever/inbox/forwarding.py [hunk 14]
```diff
@@ -152,6 +155,13 @@ def updateFieldsFromSchemata(self):
         super(ForwardingEditForm, self).updateFieldsFromSchemata()
         _drop_empty_additional_fieldset(self.groups)
 
+    def applyChanges(self, data):
+        """Records reassign activity when the responsible has changed.
+        Also update the responsible_cliend and responsible user
```
Comment (phgross) [inline opengever/inbox/forwarding.py]: responsible_cliend -> responsible_client

File: opengever/task/browser/resources/opengever.task.responsible.js [hunk 15]
```diff
@@ -1,65 +0,0 @@
-$(window).load(function() {
```
Comment (phgross) [inline opengever/task/browser/resources/opengever.task.responsible.js]: 🎉

File: sources.cfg [hunk 16]
```diff
@@ -11,3 +11,6 @@ auto-checkout = ${buildout:development-packages}
 
 [sources]
 plone.formwidget.autocomplete = git git@git.4teamwork.ch:opengever/plone.formwidget.autocomplete.git  branch=${branches:plone.formwidget.autocomplete}
+
+[branches]
+plonetheme.teamraum = mle-keyword-responsible
```
Comment (phgross) [inline sources.cfg]: As I said, please make the plonetheme.teamraum PR and remove the branch configuration.

File: opengever/ogds/base/sources.py [hunk 17]
```diff
@@ -163,3 +163,40 @@ class AllUsersAndInboxesSourceBinder(object):
...
+@implementer(IQuerySource)
+class ForwardingResponsibleSource(AllUsersAndInboxesSource):
+    """Return only users of the current orgunit and all inboxes"""
+
+    @property
+    def base_query(self):
+        query = super(ForwardingResponsibleSource, self).base_query
+        return query.filter(OrgUnit.unit_id == self.client_id)
+
+    # def search(self, query_string):
```
Comment (phgross) [inline opengever/ogds/base/sources.py]: ✂️

File: docs/HISTORY.txt (hunk 6 above)
Comment (maethu) [inline docs/HISTORY.txt]: 👍

File: opengever/ogds/base/sources.py (hunk 17 above)
Comment (maethu) [inline opengever/ogds/base/sources.py]: 👍

File: opengever/ogds/base/sources.py (hunk 10 above)
Comment (maethu) [inline opengever/ogds/base/sources.py]: 👍

File: sources.cfg (hunk 9 above)
Comment (maethu) [inline sources.cfg]: @phgross --> https://github.com/4teamwork/plonetheme.teamraum/pull/550

File: sources.cfg (hunk 16 above)
Comment (maethu) [inline sources.cfg]: 👍

File: opengever/inbox/forwarding.py (hunk 14 above)
Comment (maethu) [inline opengever/inbox/forwarding.py]: 👍

File: opengever/task/browser/resources/opengever.task.responsible.js (hunk 15 above)
Comment (maethu) [inline opengever/task/browser/resources/opengever.task.responsible.js]: 😃

**Comment by maethu** [comment]
@phgross I made the requested changes and rebased onto master.

**Comment by maethu** [comment]
@phgross it's green

**Pull request merged**